packages/react-graphql/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,198 @@
+): QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent: AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryDocumentResolving<Data, Variables> | QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options: QueryHookOptions<Data, Variables> = {},
+): QueryHookResult<Data, Variables> | QueryDocumentResolving<Data, Variables> {
+  const {
+    skip = false,
+    pollInterval,
+    notifyOnNetworkStatusChange = false,
+    client: overrideClient,
+    context,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+  } = options;
+
+  const [responseId, setResponseId] = useState(0);
+  const client = useApolloClient(overrideClient);
+  const query = useGraphQLDocument(queryOrComponent);
+
+  const watchQueryOptions: WatchQueryOptions<Variables> = useMemo(
+    () =>
+      compact({
+        context,
+        errorPolicy,
+        fetchPolicy,
+        notifyOnNetworkStatusChange,
+        pollInterval,
+        query,
+        variables,
+      }),
+    [
+      query,
+      pollInterval,
+      notifyOnNetworkStatusChange,
+      context && objectToKey(context),
+      variables && objectToKey(variables),
+      fetchPolicy,
+      errorPolicy,
+    ],
+  );
+
+  const observableQuery = useMemo(
+    () => {
+      if (!query) {
+        return;
+      }
+
+      // eslint-disable-next-line consistent-return
+      return getCachedObservableQuery<Data, Variables>(
+        client,
+        watchQueryOptions,
+      );
+    },
+    [query, client, watchQueryOptions],
+  );
+
+  useEffect(
+    () => {
+      if (skip || !observableQuery) {
+        return;
+      }
+
+      const invalidateCurrentResult = () => {
+        // A hack to get rid React warnings during tests. The default
+        // implementation of `actHack` just invokes the callback immediately.
+        // In tests, it's replaced with `act` from react-testing-library.
+        // A better solution welcome.
+        actHack(() => {
+          setResponseId(x => x + 1);
+        });
+      };
+      const subscription = observableQuery.subscribe(
+        invalidateCurrentResult,
+        invalidateCurrentResult,
+      );
+
+      invalidateCachedObservableQuery(client, watchQueryOptions);
+
+      // eslint-disable-next-line consistent-return
+      return () => {
+        subscription.unsubscribe();
+      };
+    },
+    [skip, observableQuery],
+  );
+
+  const currentResult = useMemo<
+    QueryHookResult<Data, Variables> | QueryDocumentResolving<Data, Variables>
+  >(
+    () => {
+      if (!observableQuery) {
+        return {
+          loading: true,
+          networkStatus: undefined,
```

Comment:
_carry over things that's not resolved yet but lost due to moving files_

https://github.com/Shopify/quilt/pull/642#discussion_r273724153
> I feel like this result should be basically indistinguishable from a resolved query in the loading state (so, has all the same methods; it's fine of the methods are noops)